File: huc.py
"""Watershed membership for each reservoir.

Three facts get attached to every published record: which six-digit hydrologic
unit its water drains through, whether its provider point is in Utah, and
whether its waterbody intersects Utah. They are separate on purpose. A
drainage area does not stop at a state line -- Lake
Powell's water comes down the Green from Wyoming and the Colorado from
Colorado -- so "reservoirs in Utah" and "reservoirs in drainage areas that
touch Utah" are two different questions and the dashboard has to be able to
ask each one without the other quietly changing.

Standard library only, and deliberately kept out of refresh_reservoirs.py:
this is geometry with no series, no network and no pandas in it, so it can be
tested on its own and reused by tools/probe_huc_points.py without dragging
the whole data stack along.

Boundaries come from the file the drawn scope names, written by
tools/fetch_watershed_scope.py from the USGS Watershed Boundary Dataset and
committed. Committed rather than fetched at refresh time for the same reason
as capacities.json: an assignment that can change underneath you is not
reproducible, and a reservoir that silently moves basin between two runs is
the kind of error nobody would catch by looking.

Which file that is comes from `watershed_scopes.DEFAULT_SCOPE` rather than
being written here. It was `huc6.geojson` for as long as there was only one
answer; publishing the west made it a second copy of a product decision, and
the failure mode of a second copy is that the pipeline assigns reservoirs
with one geography while every surface draws another.
"""

import json
import math
from pathlib import Path

import watershed_scopes

BOUNDARY_PATH = (watershed_scopes.ROOT
                 / watershed_scopes.get_scope(watershed_scopes.DEFAULT_SCOPE).output)
UTAH_BOUNDARY_PATH = Path(__file__).resolve().parent / "utah-boundary.geojson"

# A provider point sits in exactly one state; a waterbody need not. These were
# reviewed against the official USGS NHDPlus HR NHDWaterbody layer, and the
# permanent identifiers make the evidence reproducible without adding a remote
# geometry dependency to the daily refresh. See ADR-013 and ADR-060.
# Source: https://hydro.nationalmap.gov/arcgis/rest/services/NHDPlus_HR/MapServer/9
#
# The value is *every* state the waterbody touches, including the one holding
# the provider point, so an entry describes the waterbody rather than a
# correction to somewhere else. Bear Lake's point is in Idaho and Meeks
# Cabin's in Wyoming; both reach into Utah, which is why they are here.
#
# Absence means "the waterbody is in the state its point is in", which is true
# of every reservoir nobody has had reason to review. That is a default, not a
# finding, and `waterbody_states` says so by returning the point's state alone.
# Lake Powell carries different evidence, and it is this project's own: its
# reviewed dam point is in Coconino County, Arizona and its published
# waterbody point in San Juan County, Utah (ADR-057, ADR-058), so the water
# between them crosses the line and no external lookup is needed to say so.
# Measured across every reservoir holding both points, it is the only one --
# which is why the Utah-only table never needed it: that table existed to add
# Utah to waterbodies whose *point* was elsewhere, and Powell's point is
# already in Utah. Generalising the question exposed the gap.
#
# Keyed by the station id the roster is keyed by -- a RISE item id or an AWDB
# station triplet, exactly as `RESERVOIRS` and `AWDB_RESERVOIRS` in
# refresh_reservoirs.py are (ADR-066). A review is a fact about one reservoir,
# and a name can be shared by two: the west holds a Lost Creek in Utah and
# another in Oregon. The name stays in the value for the human reader.
CROSS_BORDER_WATERBODIES = {
    "10055500:ID:BOR": {"name": "Bear Lake", "states": ("ID", "UT"),
                        "nhd_permanent_id": "120026431"},
    "509": {"name": "Lake Powell", "states": ("AZ", "UT"),
            "evidence": "dam point in Arizona, waterbody point in Utah"},
    # Measured against the NHD polygon: 66.7% Nevada, 33.2% Arizona. RISE's
    # own five monitoring points on the lake are all in Clark County, Nevada,
    # so the provider's evidence alone would have defaulted this to Nevada and
    # been a third wrong.
    "6124": {"name": "Lake Mead", "states": ("AZ", "NV"),
             "nhd_permanent_id": "122648503"},
    "574": {"name": "Meeks Cabin", "states": ("UT", "WY"),
            "nhd_permanent_id": "120025290"},
}

def coarser_roster(codes, level: int) -> list[dict]:
    """The areas at `level` that a set of finer codes belongs to, named.

    The codes need nothing published: they are the first `level` digits of a
    code every record already carries, because HUC codes are fixed-width
    (ADR-050). The *names* have to come from somewhere, and this is ADR-048's
    rule -- the roster, not the polygons -- applied one level up.

    Read from the committed scope file for that level, which covers every area
    any scope here can reach. Absent, the names are empty and a caller labels
    by code: a filter that says "1401" is worse than one that says "Colorado
    Headwaters" and much better than no filter at all.

    Published in each surface's own payload rather than in `reference.json`,
    because every surface fetches its payload and only the maps fetch the
    reference -- and one copy of a roster is the point of having a roster.

    Lives here rather than in either refresh script because both of them
    publish it: the storage payload names the coarser areas its reservoirs
    fall in, and the snow payload names the ones its sites do (ADR-064).

    Written for one level and now taking it as an argument (ADR-073). Every
    offered level below the payload's own needs this, and a second near-copy
    of it for regions would be a second place for the fallback, the warning
    and the sort to drift.
    """
    import watershed_scopes

    scope_name = watershed_scopes.DRAWN_SCOPES.get(level)
    if scope_name is None:
        raise ValueError(
            f"no drawn scope at hydrologic level {level}; "
            f"choose {', '.join(str(k) for k in sorted(watershed_scopes.DRAWN_SCOPES))}")
    field = watershed_scopes.huc_field(level)
    names: dict[str, str] = {}
    path = watershed_scopes.ROOT / watershed_scopes.get_scope(scope_name).output
    if path.exists():
        payload = json.loads(path.read_text(encoding="utf-8"))
        for feature in payload.get("features") or []:
            code = feature["properties"].get(field)
            if code:
                names[code] = feature["properties"].get("name", "")
    else:
        print(f"WARNING: {path.name} is absent; publishing {field} codes "
              "without names")
    return [{field: code, "name": names.get(code, "")}
            for code in sorted({str(c)[:level] for c in codes if c})]


def subbasin_roster(codes) -> list[dict]:
    """The eight-digit subbasins a set of `huc8` codes names (ADR-103).

    Unlike the two coarser rosters this one cannot be derived from `huc6`:
    a finer code is not a prefix of a coarser one, so the codes handed in
    are each record's own `huc8` assignment. Same names file, same shape.
    """
    return coarser_roster(codes, 8)


def subregion_roster(codes) -> list[dict]:
    """The HUC-4 subregions a set of finer codes belongs to, named."""
    return coarser_roster(codes, 4)


def region_roster(codes) -> list[dict]:
    """The HUC-2 regions a set of finer codes belongs to, named."""
    return coarser_roster(codes, 2)


def _load_utah_polygons(path: Path = UTAH_BOUNDARY_PATH):
    """Read the committed UGRC polygon in the same normalized shape as WBD."""
    payload = json.loads(path.read_text(encoding="utf-8"))
    geometry = payload["features"][0]["geometry"]
    if geometry["type"] == "Polygon":
        return [geometry["coordinates"]]
    if geometry["type"] == "MultiPolygon":
        return geometry["coordinates"]
    raise ValueError(f"Unsupported Utah boundary geometry: {geometry['type']}")


UTAH_POLYGONS = _load_utah_polygons()
# Compatibility for callers that need the principal outline. Classification
# uses every polygon and its holes below.
UTAH_RING = UTAH_POLYGONS[0][0]

Point = tuple[float, float]

#: How far inside its drainage area an assignment point has to sit.
#:
#: The committed boundaries are generalized, so a point nearer the divide than
#: the generalization can resolve is not assigned to a basin -- it is assigned
#: to whichever side the simplification happened to leave it on.
#:
#: ADR-013 assigns from the dam because that is where the stored water leaves.
#: The rule assumes the dam is *inside* the basin, and for a dam that defines
#: the basin's own outlet it is degenerate: Hoover Dam is 0.00 km from the
#: 150100 divide, because 150100 ends at Hoover Dam. Measured across every
#: committed dam point, the next closest is Lost Lake at 2.73 km, so this
#: threshold separates the degenerate case from the real ones with room on
#: both sides. `tests/test_huc.py` holds the same 2 km against the roster.
MIN_ASSIGNMENT_MARGIN_KM = 2.0


def in_ring(point: Point, ring) -> bool:
    """Ray casting, counting a crossing on the half-open edge [y0, y1).

    The same algorithm as `inRing` in src/data/huc.ts, kept in step by a
    shared set of fixtures in the two test suites. A point exactly on the
    boundary is not specified either way and is not worth defining: dam
    points do not land on watershed boundaries, and pretending otherwise
    would invent a rule nothing tests.
    """
    x, y = point
    inside = False
    for i in range(len(ring)):
        xi, yi = ring[i][0], ring[i][1]
        xj, yj = ring[i - 1][0], ring[i - 1][1]
        if (yi > y) != (yj > y) and x < (xj - xi) * (y - yi) / (yj - yi) + xi:
            inside = not inside
    return inside


def in_polygon(point: Point, rings) -> bool:
    """Inside the outer ring and inside none of its holes."""
    if not rings or not in_ring(point, rings[0]):
        return False
    return not any(in_ring(point, hole) for hole in rings[1:])


def in_utah(point: Point) -> bool:
    return any(in_polygon(point, polygon) for polygon in UTAH_POLYGONS)


def waterbody_states(station: str | None, point_state: str | None) -> list[str]:
    """Every state this reservoir's water touches.

    Looked up by the station id the roster is keyed by (ADR-066): a review is
    a fact about one reservoir, and a name can be shared by two. The reviewed
    answer where there is one, and the point's own state otherwise. Sorted so
    two records carrying the same states compare equal however the table was
    written.

    A reservoir whose point falls in no state at all -- which the mask can
    produce for a waterbody just off a generalized outline -- returns the
    reviewed list if it has one and an empty list if it does not, rather than
    inventing a state to be in.
    """
    reviewed = CROSS_BORDER_WATERBODIES.get(station)
    if reviewed:
        return sorted(reviewed["states"])
    return [point_state] if point_state else []


def waterbody_intersects_utah(station: str | None, point: Point) -> bool:
    """Whether the reservoir surface intersects Utah.

    A point inside Utah proves intersection. A point outside the state needs
    a reviewed polygon; the current exceptions are versioned above, by
    station id (ADR-066).
    """
    reviewed = CROSS_BORDER_WATERBODIES.get(station)
    return in_utah(point) or bool(reviewed and "UT" in reviewed["states"])


def location_fields(station: str | None, lat: float, lon: float) -> dict:
    """Stable location facts that do not depend on watershed boundaries."""
    site = (lon, lat)
    return {
        "in_utah": in_utah(site),
        "intersects_utah": waterbody_intersects_utah(station, site),
    }


def _bounds(polygons) -> tuple[float, float, float, float]:
    """West, south, east, north for every ring of every part."""
    west = south = float("inf")
    east = north = float("-inf")
    for polygon in polygons:
        for ring in polygon:
            for lon, lat in ring:
                if lon < west:
                    west = lon
                if lon > east:
                    east = lon
                if lat < south:
                    south = lat
                if lat > north:
                    north = lat
    return west, south, east, north


#: Decimal places a published drainage-area bounding box is rounded to.
#:
#: The source geometry is committed at five (`GEOMETRY_PRECISION` in
#: `tools/fetch_watershed_scope.py`), and a box does not need that much: it
#: only has to say roughly where a map should open, not trace a ring, and
#: every extra decimal is bytes every map page fetches on 75+44+14+10+5
#: units. Three is about 111 metres at the equator -- the same 0.001-degree
#: margin `extent.test.ts` already allows a freshly measured box to differ
#: from the frozen module's `HUC6_BOUNDS` by.
PUBLISHED_BBOX_DECIMALS = 3


def outer_bbox(
    bounds: tuple[float, float, float, float],
    decimals: int = PUBLISHED_BBOX_DECIMALS
) -> list[float]:
    """``[west, south, east, north]``, coarsened so the box still contains
    the rings ``bounds`` was measured from.

    The two edges cannot round the same way. Rounding the minimum up or the
    maximum down moves that edge *inward* -- toward the middle of the box --
    which is exactly the direction that clips the polygon the box is
    published to describe. So the minima floor and the maxima ceil: rounding
    outward rather than to the nearest value is what keeps a coarsened box a
    superset of the exact one ``_bounds`` measured, not an approximation of
    it that happens to be close.
    """
    west, south, east, north = bounds
    scale = 10 ** decimals
    return [
        math.floor(west * scale) / scale,
        math.floor(south * scale) / scale,
        math.ceil(east * scale) / scale,
        math.ceil(north * scale) / scale,
    ]


def unit_code(properties: dict) -> str:
    """The unit's code, whatever size the collection is.

    Hydrologic codes are fixed width, so the level is the digit count and the
    attribute is named after it -- `huc6` in a six-digit collection, `huc8`
    in an eight. Reading a fixed `huc6` raised a KeyError against the western
    HUC-8 file, which is the polite version of this mistake; the client
    version of it parsed the payload as no areas at all and drew a blank map
    (ADR-050).
    """
    for level in (2, 4, 6, 8, 10, 12):
        code = properties.get(f"huc{level}")
        if isinstance(code, str):
            return code
    raise KeyError(
        "feature carries no huc code; expected one of "
        + ", ".join(f"huc{level}" for level in (2, 4, 6, 8, 10, 12)))


def units_from_collection(payload: dict) -> list[dict]:
    """Normalize a GeoJSON feature collection to the assignment shape.

    Most callers use :func:`load_units` with a committed file. Research tools
    also need to classify divide-adjacent points against an un-generalized
    federal response without writing that much larger geometry to the
    repository first.

    Each unit carries its bounding box, computed once here. See
    :func:`assign_huc` for why.
    """
    units = []
    for feature in payload["features"]:
        geometry = feature["geometry"]
        coordinates = geometry["coordinates"]
        polygons = (coordinates if geometry["type"] == "MultiPolygon"
                    else [coordinates])
        units.append({
            "huc6": unit_code(feature["properties"]),
            "name": feature["properties"]["name"],
            "states": feature["properties"].get("states", ""),
            "polygons": polygons,
            "bounds": _bounds(polygons),
        })
    return sorted(units, key=lambda unit: unit["huc6"])


def load_units(path: Path | None = None) -> list[dict]:
    """The committed hydrologic units, normalized to one polygon list each."""
    payload = json.loads((path or BOUNDARY_PATH).read_text())
    return units_from_collection(payload)


def load_units_at(level: int) -> list[dict]:
    """The committed units of the drawn scope at `level` (ADR-103).

    Read through `DRAWN_SCOPES` rather than a path, for ADR-063's reason:
    which file holds which geography has moved before. Each unit's code is
    under the `huc6` key whatever its width, as every unit list here is.
    """
    scope_name = watershed_scopes.DRAWN_SCOPES[level]
    return load_units(watershed_scopes.ROOT / watershed_scopes.get_scope(scope_name).output)


def assign_huc(point: Point, units) -> dict | None:
    """The unit containing this point, or None.

    First hit wins. Hydrologic units tile the country without overlapping, so
    a point inside two of them means the boundary data is wrong; picking the
    first is no worse than any other arbitrary choice, and
    tests/test_huc.py asserts the situation does not arise.

    The bounding-box test in front of the ring scan is what makes this
    affordable at western scale. Ray casting walks every vertex of every ring
    it is given: measured at 44 ms a point on the pre-scoping western HUC-8
    file (1,247 units, 815,761 vertices; the drainage-scoped file committed
    since is 571 units and roughly half the vertices), which is half a minute
    for a 690-reservoir roster and over a minute for the snow network. A box
    comparison rejects almost all of them in four float comparisons, and the
    answer is identical -- a point outside the box cannot be inside the
    polygon.

    `bounds` is optional so that a unit built by hand, in a test or an older
    caller, still works; it is simply slower.
    """
    lon, lat = point
    for unit in units:
        bounds = unit.get("bounds")
        if bounds is not None:
            west, south, east, north = bounds
            if lon < west or lon > east or lat < south or lat > north:
                continue
        if any(in_polygon(point, polygon) for polygon in unit["polygons"]):
            return unit
    return None


def distance_to_boundary_km(point: Point, unit: dict) -> float:
    """Shortest distance from the point to any edge of the unit.

    This is the number that says how much boundary precision the assignment
    needs. A reservoir 200 m from a boundary could be moved into the next
    unit by a generalized polygon or a slightly different dam coordinate; one
    20 km inside cannot. Measured across the 53 reservoirs published at the
    time of the boundary study, the closest is 2.72 km, which is what
    justifies the 500 m generalization in
    tools/fetch_watershed_scope.py.

    Computed on a local equirectangular projection about the point. Over the
    few kilometres that matter the error is far below the thing being
    measured, and it avoids a geodesic dependency in a module that is
    deliberately standard library only.
    """
    lon, lat = point
    scale = math.cos(math.radians(lat))
    km_per_degree = 111.32

    best = float("inf")
    for polygon in unit["polygons"]:
        for ring in polygon:
            local = [((vertex[0] - lon) * scale * km_per_degree,
                      (vertex[1] - lat) * km_per_degree) for vertex in ring]
            for i in range(len(local) - 1):
                (x1, y1), (x2, y2) = local[i], local[i + 1]
                dx, dy = x2 - x1, y2 - y1
                length_squared = dx * dx + dy * dy
                if length_squared == 0:
                    best = min(best, math.hypot(x1, y1))
                    continue
                # Projection of the origin onto the segment, clamped to it.
                t = max(0.0, min(1.0, -(x1 * dx + y1 * dy) / length_squared))
                best = min(best, math.hypot(x1 + t * dx, y1 + t * dy))
    return best


def haversine_km(a: Point, b: Point) -> float:
    lon1, lat1, lon2, lat2 = map(math.radians, (a[0], a[1], b[0], b[1]))
    h = (math.sin((lat2 - lat1) / 2) ** 2
         + math.cos(lat1) * math.cos(lat2) * math.sin((lon2 - lon1) / 2) ** 2)
    return 2 * 6371.0088 * math.asin(math.sqrt(h))


def describe(lat: float, lon: float, units, *, station: str | None,
             assignment_point: Point | None = None,
             source: str = "published_point",
             fine_units=None) -> dict:
    """The watershed fields for one reservoir record.

    `station` is the id the roster keys the reservoir by (ADR-066); it is
    what the cross-border waterbody review is looked up with.

    Two points, and they must not be collapsed into one:

    - The **reservoir's** point decides `in_utah`. That preserves the provider
      point-location fact.
    - The reservoir point plus reviewed waterbody polygons decide
      `intersects_utah`. That owns the default Utah scope.
    - The **assignment** point decides the drainage area. That asks where the
      stored water leaves, which is the dam or outlet.

    They were the same in the original assignment study and will not stay
    that way. Glen Canyon Dam is in Arizona while Lake Powell reaches well
    into Utah, so the moment the dam points land, computing `in_utah` from the
    assignment point would drop the single largest reservoir on this
    dashboard out of its own default view.

    `fine_units` are the subbasins (ADR-103). The same point that chose the
    basin chooses the subbasin, divide fallback included, so the two codes
    cannot disagree about which side of a ridge the water leaves on; the
    subbasin's code is published beside the basin's, never in its place.

    `source` records what kind of point produced the assignment, because the
    answer is going to improve. The published coordinates are lake points, a
    median of 1.08 km from the dam. Across the 53 reservoirs published when
    that study was run, using the dam point instead moved none of them, so the
    upgrade was a correctness improvement rather than a correction -- but a
    reader should still be able to tell which one produced a given row.
    """
    site = (lon, lat)
    point = tuple(assignment_point) if assignment_point else site
    unit = assign_huc(point, units)
    # A dam sitting on the divide it defines cannot say which side it is on.
    # Fall back to the waterbody, which is unambiguously upstream of it, and
    # record that the fallback happened rather than quietly taking it.
    if (unit is not None and assignment_point is not None
            and distance_to_boundary_km(point, unit) < MIN_ASSIGNMENT_MARGIN_KM):
        from_site = assign_huc(site, units)
        if (from_site is not None
                and distance_to_boundary_km(site, from_site) >= MIN_ASSIGNMENT_MARGIN_KM):
            point, unit = site, from_site
            source = "published_point_dam_on_divide"
    fine = assign_huc(point, fine_units) if fine_units else None
    return {
        **location_fields(station, lat, lon),
        "huc6": unit["huc6"] if unit else None,
        "huc6_name": unit["name"] if unit else None,
        "huc8": fine["huc6"] if fine else None,
        "huc8_name": fine["name"] if fine else None,
        # Every state the drainage area reaches, which is a different question
        # from where the reservoir is (ADR-060). Lake Powell sits in Utah and
        # its water arrives from Wyoming and Colorado; a reader asking "what
        # does Colorado feed" wants this list, and one asking "what is in
        # Colorado" wants the state above. The drainage file already carries
        # it, so this costs a split rather than a lookup.
        "connected_states": sorted(
            s for s in (unit.get("states") or "").split(",") if s) if unit else [],
        "huc_assignment_point": [round(point[0], 5), round(point[1], 5)],
        "huc_assignment_source": source if unit else None,
    }

File: src/data/huc.ts
/*
 * Watershed (HUC6) assignment and capacity-weighted rollups.
 *
 * Phase 1.5 of docs/history/modernization-2026.md. Pure arithmetic and geometry with no
 * DOM, no network and no SDK in it, so all of it is unit-testable -- which
 * is the point: the map pages already draw the HUC6 boundaries live from the
 * USGS service, but nothing yet says which unit a reservoir belongs to, and
 * a wrong assignment is invisible on a map.
 *
 * Two rules from the plan are load-bearing and are enforced here rather than
 * left to callers:
 *
 *   1. A reservoir is assigned by its dam or outlet point. A large reservoir
 *      can span two units; the assignment answers "where does the stored
 *      water leave?", which is one place.
 *   2. A twelve-month HUC value is valid only when *every* tracked reservoir
 *      in that unit has a value for that month. Otherwise the series shows a
 *      gap. This deliberately differs from `statewideMonthly` in
 *      shared/reservoir-viz.js, which sums whatever is present: across 53
 *      reservoirs one missing month barely moves the total, but a unit with
 *      three reservoirs would show a cliff and read as a drought.
 */

import type { MonthlyRecord, NullableNumber } from "../types";

/** Longitude, latitude -- GeoJSON order, not lat/lon. */
/**
 * The shape of a hydrologic unit code, at any level this project reads.
 *
 * Codes are fixed-width and zero-padded, and the levels are the even numbers
 * from 2 to 12 -- a region is two digits, a subregion four, a basin six, and
 * so on down. So the test is "an even count of digits, at most twelve", which
 * also rejects the five- and seven-digit strings that are not a level at all
 * and would otherwise be waved through by a looser `\d+`.
 *
 * One constant because four places were each carrying their own `/^\d{6}$/`,
 * and every one of them would have had to be found again the first time a
 * payload arrived at another level.
 */
export const HUC_CODE = /^(?:\d{2}){1,6}$/;

/**
 * A record's drainage-area code at a drawn level (ADR-103).
 *
 * Coarser than the basin is a slice, because fixed-width codes nest
 * (ADR-050). Finer than the basin is the record's own `huc8` -- a subbasin
 * is a prefix of nothing, so it cannot be sliced out of `huc6` -- and null
 * when the record carries none, so a caller leaves it out rather than
 * filing a six-digit code among eight-digit ones.
 */
export function drainageCodeAtLevel(
  huc6: string | null | undefined,
  huc8: string | null | undefined,
  level: number
): string | null {
  if (level >= 8) return typeof huc8 === "string" && huc8.length === 8 ? huc8 : null;
  if (typeof huc6 !== "string" || huc6.length < level) return null;
  return huc6.slice(0, level);
}

export type Point = readonly [number, number];

/** A linear ring: the outer ring first, then any holes. */
export type Ring = readonly Point[];

export interface HucUnit {
  huc6: string;
  name: string;
  /** Comma-separated state codes as the USGS service publishes them. */
  states: string;
  /** One polygon is one ring list; a multipolygon is several. */
  polygons: readonly (readonly Ring[])[];
}

/**
 * The fields a rollup needs. Declared structurally rather than as
 * `Reservoir` so tests can build three-line fixtures, and so a future
 * candidate site that is not yet a published reservoir can be rolled up too.
 */
export interface HucMember {
  name: string;
  huc6?: string | null;
  huc6_name?: string | null;
  current_storage_af: number;
  capacity_af?: NullableNumber;
  record_max_af: number;
  monthly?: readonly MonthlyRecord[];
}

export interface HucRollup {
  huc6: string;
  name: string;
  /** Reservoirs counted, after removing duplicates. */
  count: number;
  storageAf: number;
  capacityAf: number;
  /** Capacity-weighted: sum(storage) / sum(capacity). Null with no denominator. */
  percentFull: number | null;
  /** Reservoirs with no capacity and no observed maximum to stand in for one. */
  withoutCapacity: number;
  members: readonly string[];
}

export interface HucMonthlyPoint {
  month: string;
  /** Null when the unit is not fully covered for this month. */
  meanAf: number | null;
  normalAf: number | null;
  /** How many of the unit's reservoirs reported this month. */
  covered: number;
  count: number;
}

export interface CoverageRow {
  name: string;
  huc6: string | null;
  result: "assigned" | "unassigned";
  reason: string;
}

const normalizeName = (name: string): string => name.trim().toLowerCase();

/**
 * Ray casting, counting a crossing on the half-open edge [y0, y1) so a
 * vertex is not counted twice. A point on the boundary is not specified
 * either way and is not worth defining: dam points do not land on watershed
 * boundaries, and pretending otherwise would invent a rule nothing tests.
 */
function inRing(point: Point, ring: Ring): boolean {
  const [x, y] = point;
  let inside = false;
  for (let i = 0, j = ring.length - 1; i < ring.length; j = i++) {
    const from = ring[j];
    const to = ring[i];
    if (from === undefined || to === undefined) continue;
    const [xi, yi] = to;
    const [xj, yj] = from;
    if ((yi > y) !== (yj > y) && x < (xj - xi) * (y - yi) / (yj - yi) + xi) {
      inside = !inside;
    }
  }
  return inside;
}

function inPolygon(point: Point, rings: readonly Ring[]): boolean {
  const [outer, ...holes] = rings;
  if (outer === undefined || !inRing(point, outer)) return false;
  return !holes.some((hole) => inRing(point, hole));
}

/**
 * The unit containing this point, or null. Units are tested in order and the
 * first hit wins: HUC6 units tile the country without overlapping, so a
 * point in two of them means the boundary data is wrong, and picking the
 * first is no worse than any other arbitrary choice.
 */
export function assignHuc(point: Point, units: readonly HucUnit[]): HucUnit | null {
  return units.find((unit) =>
    unit.polygons.some((polygon) => inPolygon(point, polygon))) ?? null;
}

/**
 * The same capacity fallback the rest of the app uses: real capacity where
 * we have it, otherwise the highest storage observed since 2015. A unit
 * whose reservoirs have neither contributes nothing to the denominator, and
 * says so through `withoutCapacity`.
 */
export function capacityBasis(member: HucMember): number {
  return member.capacity_af ?? member.record_max_af;
}

/**
 * Removes duplicate sites by name. A reservoir can arrive twice when a
 * candidate list from the Reclamation facility service is merged with the
 * published inventory; counting it twice would inflate both the storage and
 * the capacity of its unit.
 */
function dedupe(members: readonly HucMember[]): HucMember[] {
  const seen = new Set<string>();
  return members.filter((member) => {
    const key = normalizeName(member.name);
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}

/** Members grouped by unit, in first-seen order, skipping the unassigned. */
function groupByHuc(members: readonly HucMember[]): Map<string, HucMember[]> {
  const groups = new Map<string, HucMember[]>();
  for (const member of dedupe(members)) {
    const huc6 = member.huc6;
    if (huc6 === null || huc6 === undefined || huc6 === "") continue;
    const group = groups.get(huc6);
    if (group === undefined) groups.set(huc6, [member]);
    else group.push(member);
  }
  return groups;
}

/**
 * One capacity-weighted total per unit. Label the result "tracked reservoir
 * storage in this drainage area" -- it is not the percentage of all water in
 * the watershed, which is why every row carries its reservoir count and
 * combined capacity.
 */
export function rollupByHuc(members: readonly HucMember[]): HucRollup[] {
  return [...groupByHuc(members)].map(([huc6, group]) => {
    const storageAf = group.reduce((total, member) =>
      total + member.current_storage_af, 0);
    const capacityAf = group.reduce((total, member) =>
      total + Math.max(capacityBasis(member), 0), 0);
    return {
      huc6,
      name: group.find((member) => member.huc6_name)?.huc6_name ?? huc6,
      count: group.length,
      storageAf,
      capacityAf,
      percentFull: capacityAf > 0 ? storageAf / capacityAf * 100 : null,
      withoutCapacity: group.filter((member) => capacityBasis(member) <= 0).length,
      members: group.map((member) => member.name)
    };
  }).sort((a, b) => b.capacityAf - a.capacityAf);
}

/**
 * Twelve months per unit, gated on full coverage. Months are the union of
 * every member's months, so a unit whose reservoirs report different spans
 * shows the gaps rather than silently shortening its own history.
 */
export function monthlyRollupByHuc(
  members: readonly HucMember[]
): Map<string, HucMonthlyPoint[]> {
  const out = new Map<string, HucMonthlyPoint[]>();
  for (const [huc6, group] of groupByHuc(members)) {
    const months = [...new Set(group.flatMap((member) =>
      (member.monthly ?? []).map((record) => record.month)))].sort();
    out.set(huc6, months.map((month) => {
      const values = group.map((member) =>
        (member.monthly ?? []).find((record) => record.month === month));
      const present = values.filter((record) =>
        record !== undefined && record.mean_af !== null);
      const covered = present.length;
      const complete = covered === group.length;
      const normals = present.map((record) => record?.normal_af ?? null);
      return {
        month,
        meanAf: complete
          ? present.reduce((total, record) => total + (record?.mean_af ?? 0), 0)
          : null,
        normalAf: complete && normals.every((value) => value !== null)
          ? normals.reduce((total: number, value) => total + (value ?? 0), 0)
          : null,
        covered,
        count: group.length
      };
    }));
  }
  return out;
}

/**
 * One row per site, saying what happened to it and why. The plan asks for a
 * published coverage report; this is the data behind it. An unassigned site
 * is not a failure to hide -- it is the thing a reviewer needs to see.
 */
export function coverageReport(members: readonly HucMember[]): CoverageRow[] {
  return dedupe(members).map((member) => {
    const huc6 = member.huc6 ?? null;
    if (huc6 === null || huc6 === "") {
      return {
        name: member.name,
        huc6: null,
        result: "unassigned" as const,
        reason: "no dam or outlet point fell inside a published unit"
      };
    }
    return {
      name: member.name,
      huc6,
      result: "assigned" as const,
      reason: capacityBasis(member) > 0
        ? "assigned by its dam or outlet point"
        : "assigned, but it adds no capacity to the unit total"
    };
  });
}
